src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java

```diff
@@ -151,24 +163,35 @@ public ProductReferenceResolver(@Nonnull final ProductSyncOptions productSyncOpt
             });
     }
 
+    /**
+     * Given a {@link ProductDraftBuilder} this method attempts to resolve the categories and categoryOrderHints to
+     * return a {@link CompletionStage} which contains a new instance of the builder with the resolved references.
+     * The key of the category references is either taken from the expanded references or taken from the value of the
+     * id fields.
+     *
+     * @param draftBuilder the productDraft to resolve its category and categoryOrderHints references.
+     * @return a {@link CompletionStage} that contains as a result a new builder instance with resolved references or,
+     *         in case an error occurs during reference resolution, a {@link ReferenceResolutionException}.
+     */
     @Nonnull
-    private CompletionStage<ProductDraftBuilder> resolveCategoryReferences(
-            @Nonnull final ProductDraftBuilder draftBuilder) {
+    public CompletionStage<ProductDraftBuilder> resolveCategoryReferences(
+        @Nonnull final ProductDraftBuilder draftBuilder) {
         final Set<ResourceIdentifier<Category>> categoryResourceIdentifiers = draftBuilder.getCategories();
         final Set<String> categoryKeys = new HashSet<>();
-
-        categoryResourceIdentifiers.forEach(categoryResourceIdentifier -> {
+        for (ResourceIdentifier<Category> categoryResourceIdentifier: categoryResourceIdentifiers) {
             if (categoryResourceIdentifier != null) {
                 try {
                     final String categoryKey = getKeyFromResourceIdentifier(categoryResourceIdentifier,
                         options.shouldAllowUuidKeys());
                     categoryKeys.add(categoryKey);
                 } catch (ReferenceResolutionException referenceResolutionException) {
-                    options.applyErrorCallback(format(FAILED_TO_RESOLVE_CATEGORY, draftBuilder.getKey(),
-                        referenceResolutionException), referenceResolutionException);
+                    return exceptionallyCompletedFuture(
```

Comment:
Because if one category's reference is not resolved correctly, the product won't be able to be created/updated anyways, so it doesn't make sense to continue resolving the rest of the category references.